graph/src/components/store.rs

```diff
@@ -557,6 +556,60 @@ impl EntityOperation {
     }
 }
 
+/// An operation on subgraph metadata. All operations implictly only concern
+/// the subgraph of subgraphs.
+#[derive(Clone, Debug, PartialEq)]
+pub enum MetadataOperation {
+    /// Locates the entity with type `entity` and the given `id` in the
+    /// subgraph of subgraphs and sets its attributes according to the
+    /// contents of `data`.  If no such entity exists, creates a new entity.
+    Set {
+        entity: String,
+        id: String,
+        data: Entity,
+    },
+
+    /// Removes an entity with the specified entity type and id if one exists.
+    Remove { entity: String, id: String },
+
+    /// Aborts and rolls back the transaction unless `query` returns entities
+    /// exactly matching `entity_ids`.  The equality test is only sensitive
+    /// to the order of the results if `query` contains an `order_by`.
+    AbortUnless {
+        description: String, // Programmer-friendly debug message to explain reason for abort
+        query: EntityQuery,  // The query to run
+        entity_ids: Vec<String>, // What entities the query should return
+    },
+
+    /// Update an entity. The `data` should only contain the attributes that
+    /// need to be changed, not the entire entity. The update will only happen
+    /// if the given entity matches `guard` when the update is made. `Update`
+    /// provides a way to atomically do a check-and-set change to an entity.
+    Update {
+        entity: String,
+        id: String,
+        data: Entity,
+        guard: Option<EntityFilter>,
+    },
+}
+
+impl MetadataOperation {
+    pub fn entity_key(&self) -> Option<EntityKey> {
```

Comment:
No, we only `unwrap` when we know it is safe to do so; this is correct to me as an API: it's the caller's responsibility to figure out how it wants to handle the option.

Reply:
What about avoiding any unwrapping at all by having this be `fn entity_key(entity: String, id: String) -> EntityKey`?

Reply:
Is your objection that the `unwrap` looks like it might panic or something else? The point of this method is to get the key from the `MetadataOperation` if it has one (i.e., it's not an `AbortUnless`). And only in those cases do we have an `entity` and a `id` embedded in the operation.

Reply:
Yes, unwrap looks like it might panic, so it always makes you pause when reading the code to convince yourself that it's correct. In some cases I think it's worth it to simplify the code, in this case I think we can easily avoid any unwraps.

Reply:
Addressed in commit 676b22da